plugins/traefik/main.go

```diff
@@ -136,11 +164,132 @@ func (r *responseWriter) Hijack() (net.Conn, *bufio.ReadWriter, error) {
 	if !ok {
 		return nil, nil, fmt.Errorf("%T is not a http.Hijacker", r.responseWriter)
 	}
-	return hijacker.Hijack()
+	if r.websocket {
+		fmt.Println("=== hijack for websocket")
+		conn, bufio, err := hijacker.Hijack()
+		return newConnWrapper(conn), bufio, err
+	} else {
+		return hijacker.Hijack()
+	}
 }
 
 func (r *responseWriter) Flush() {
 	if flusher, ok := r.responseWriter.(http.Flusher); ok {
 		flusher.Flush()
 	}
 }
+
+func isWebsocketRequest(req *http.Request) bool {
+	return containsHeader(req, "Connection", "upgrade") && containsHeader(req, "Upgrade", "websocket")
+}
+
+func containsHeader(req *http.Request, name, value string) bool {
+	items := strings.Split(req.Header.Get(name), ",")
+	for _, item := range items {
+		if value == strings.ToLower(strings.TrimSpace(item)) {
+			return true
+		}
+	}
+	return false
+}
+
+func newConnWrapper(c net.Conn) *conn {
+	return &conn{
+		conn: c,
+	}
+}
+
+type conn struct {
+	conn net.Conn
+}
+
+// LocalAddr implements net.Conn.
+func (c *conn) LocalAddr() net.Addr {
+	panic("unimplemented")
+}
+
+// RemoteAddr implements net.Conn.
+func (c *conn) RemoteAddr() net.Addr {
+	panic("unimplemented")
+}
+
+// SetDeadline implements net.Conn.
+func (c *conn) SetDeadline(t time.Time) error {
+	panic("unimplemented")
+}
+
+// SetReadDeadline implements net.Conn.
+func (c *conn) SetReadDeadline(t time.Time) error {
+	panic("unimplemented")
+}
+
+// SetWriteDeadline implements net.Conn.
+func (c *conn) SetWriteDeadline(t time.Time) error {
+	panic("unimplemented")
+}
+
+func (c *conn) Read(b []byte) (n int, err error) {
+	n, err = c.conn.Read(b)
+	if err == nil {
+		wsEventChan <- WebSocketRead // Notify about the read operation
+	}
+	return
+}
+
+func (c *conn) Write(b []byte) (n int, err error) {
+	n, err = c.conn.Write(b)
+	if err == nil {
+		wsEventChan <- WebSocketWrite // Notify about the write operation
+	}
+	return
+}
+
+func (c *conn) Close() error {
+	err := c.conn.Close()
+	wsEventChan <- WebSocketClose // Notify about the close operation
+	return err
+}
+
+func monitorWebSocketActivity(sablierRequest *http.Request, sm *SablierMiddleware) {
+	duration, err := time.ParseDuration(sm.config.SessionDuration)
+	if err != nil {
+		fmt.Fprintln(os.Stdout, []any{`Error parsing sessionDuration: %v`, err}...)
+		return
+	}
+	alertTime := duration - (duration * 5 / 100) // Calculate alert time at 95% of the total duration
+	alertTicker := time.NewTicker(alertTime)
+	defer alertTicker.Stop()
+
+	// Active flag to determine if the ticker should be reset
+	activeDuringAlert := false
+
+	for {
+		select {
+		case event := <-wsEventChan:
+			switch event {
+			case WebSocketRead, WebSocketWrite:
+				activeDuringAlert = true // Mark that there was activity during the alert period
+
+			case WebSocketClose:
+				fmt.Println("WebSocket closed")
+				activeDuringAlert = false // Do not reset ticker on close
```

Comment:
you can complete go routine here. No need to wait for ticker.

Reply:
check my comment above